### DISCUSSION TITLE
why _finalize_fairy not raise Exception?

### DISCUSSION THREAD

**Discussion by 522623905** [sqlalchemy/sqlalchemy#7999] [Usage Questions]
sqlalchemy/pool.py:_finalize_fairy

            # Immediately close detached instances
            if not connection_record:
                if pool.dispatch.close_detached:
                    pool.dispatch.close_detached(connection)
                pool._close_connection(connection)
        except BaseException as e:
            pool.logger.error(
                "Exception during reset or similar", exc_info=True)
            if connection_record:
                connection_record.invalidate(e=e)
            if not isinstance(e, Exception):
                raise

Hello, I want to know, why is it judged to be an exception here and not throw an exception?

**Comment by CaselIT** [comment]
Hi,

`_finalize_fairy` is a very internal function that should never be used by user code. What's the contex of your question?

**Reply by 522623905** [reply]
this bt ：
-> resource = query.one_or_none()
/usr/lib/python2.7/site-packages/sqlalchemy/orm/query.py(2784)one_or_none()
-> ret = list(self)
  /usr/lib/python2.7/site-packages/sqlalchemy/orm/loading.py(71)instances()
-> fetch = cursor.fetchall()
  /usr/lib/python2.7/site-packages/sqlalchemy/engine/result.py(1120)fetchall()
-> self._soft_close()
  /usr/lib/python2.7/site-packages/sqlalchemy/engine/result.py(824)_soft_close()
-> self.connection.close()
  /usr/lib/python2.7/site-packages/sqlalchemy/engine/base.py(859)close()
-> conn.close()
  /usr/lib/python2.7/site-packages/sqlalchemy/pool.py(963)close()
-> self._checkin()
  /usr/lib/python2.7/site-packages/sqlalchemy/pool.py(814)_checkin()
-> self._pool, None, self._echo, fairy=self)
> /usr/lib/python2.7/site-packages/sqlalchemy/pool.py(701)_finalize_fairy()
-> raise

this error  is  OperationalError：
class OperationalError(DatabaseError)
class DatabaseError(Error)
class Error(MySQLError)
class MySQLError(Exception)

but  _finalize_fairy judge  is Exeception， not raise，causing the upper layer to not perceive 。
So I want to know the reason。

**Reply by zzzeek** [reply]
finalize fairy will log the exception if a close fails to proceed.  Turn on [logging](https://docs.sqlalchemy.org/en/14/core/engines.html#configuring-logging) to see exceptions generated when it closes out connections here.

**Reply by 522623905** [reply]
If the exception is thrown, will have there any problem?

**Reply by zzzeek** [reply]
not directly, that's a connection being discarded.   It's something you might want to know about to see for example what might have been wrong with that connection if it had produced an error or invalidation scenario earlier.

**Reply by 522623905** [reply]
why the connection is discarded？
with sql.session:
     xxx
     xxx
result：
failed，but not raise。However, not all scenarios do not throw。

**Reply by zzzeek** [reply]
pooled connections are closed / discarded all the time for many reasons.  impossible to know without a complete working example
